src/util-lua-sandbox.c

```diff
@@ -0,0 +1,300 @@
+LUAMOD_API int luaopen_sandbox(lua_State *L);
+
+static void *sb_alloc(void *ud, void *ptr, size_t osize, size_t nsize)
+{
+    (void)ud;
+    (void)osize; /* not used */
+    sb_lua_state *ctx = (sb_lua_state *)ud;
+    if (nsize == 0) {
+        if (ptr != NULL) {
+            // ASSERT: alloc_bytes > osize
+            ctx->alloc_bytes -= osize;
+        }
+        SCFree(ptr);
+        return NULL;
+    } else {
+        // We can be a bit sloppy on the alloc limit since it's not supposed to be hit.
+        //  ASSERT: ctx->alloc_bytes + nsize > ctx->alloc_bytes
+        if (ctx->alloc_bytes + nsize > ctx->alloc_limit) {
+            // TODO: Trace in a better way
+            return NULL;
+        }
+        void *nptr = SCRealloc(ptr, nsize);
+        if (nptr != NULL) {
+            ctx->alloc_bytes += nsize;
+        }
+        return nptr;
+    }
+}
+
+/*
+** These are the set of libs allowed in the restricted lua sandbox
+*/
+static const luaL_Reg sb_restrictedlibs[] = { { LUA_GNAME, luaopen_base },
+    //  {LUA_LOADLIBNAME, luaopen_package},
+    //  {LUA_COLIBNAME, luaopen_coroutine},
+    { LUA_TABLIBNAME, luaopen_table },
+    //  {LUA_IOLIBNAME, luaopen_io},
+    //  {LUA_OSLIBNAME, luaopen_os},
+    { LUA_STRLIBNAME, luaopen_string }, { LUA_MATHLIBNAME, luaopen_math },
+    { LUA_UTF8LIBNAME, luaopen_utf8 },
+    { LUA_DBLIBNAME, luaopen_sandbox }, // TODO: remove this from restricted
+    { NULL, NULL } };
+
+// TODO: should we block raw* functions?
+// TODO: Will we ever need to block a subset of functions more than one level deep?
+static const sb_block_function sb_restrictedfuncs[] = { { LUA_GNAME, "collectgarbage" },
+    { LUA_GNAME, "dofile" }, { LUA_GNAME, "getmetatable" }, { LUA_GNAME, "loadfile" },
+    { LUA_GNAME, "load" }, { LUA_GNAME, "pcall" }, { LUA_GNAME, "setmetatable" },
+    { LUA_GNAME, "xpcall" },
+
+    { LUA_STRLIBNAME, "rep" }, // TODO:  probably don't need to block this for normal restricted
+                               // since we have memory limit
+    { NULL, NULL } };
+
+static void sb_loadlibs(lua_State *L, const luaL_Reg *libs)
+{
+    const luaL_Reg *lib;
+    /* "require" functions from 'loadedlibs' and set results to global table */
+    for (lib = libs; lib->func; lib++) {
+        luaL_requiref(L, lib->name, lib->func, 1);
+        lua_pop(L, 1); /* remove lib */
+    }
+}
+
+static void sb_block_functions(lua_State *L, const sb_block_function *funcs)
+{
+    const sb_block_function *func;
+
+    // set target functions to nil
+    lua_pushglobaltable(L);
+    for (func = funcs; func->module; func++) {
+        lua_pushstring(L, func->module);
+        lua_gettable(L, -2); // load module to stack
+        lua_pushstring(L, func->name);
+        lua_pushnil(L);
+        lua_settable(L, -3);
+        lua_pop(L, 1); // remove module from the stack
+    }
+    lua_pop(L, 1); // remove global table
+}
+
+LUALIB_API void sb_loadrestricted(lua_State *L)
```

Comment:
I don't think we need these yet... Looks like they are for library support on Windows.  At some point we'll probably need to do a sweep with our own variant, which would be used here instead of LUALIB_API.

Any reason to keep these @J0eJ0h ?

Reply:
I think for Linux it is just extern "c".  No reason not to use the suricata equivalent.